tests/async/test_browsertype_connect_cdp.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Microsoft Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict
+
+import pytest
+import requests
+
+from playwright.async_api import BrowserType
+from tests.server import Server
+
+PORT = 9339
+
+
+@pytest.mark.only_browser("chromium")
+async def test_connect_to_an_existing_cdp_session(
+    launch_arguments: Dict, browser_type: BrowserType
+):
+    browser_server = await browser_type.launch(
+        **launch_arguments, args=[f"--remote-debugging-port={PORT}"]
+    )
+    cdp_browser = await browser_type.connect_over_cdp(f"http://localhost:{PORT}")
+    assert len(cdp_browser.contexts) == 1
+    await cdp_browser.close()
+    await browser_server.close()
+
+
+@pytest.mark.only_browser("chromium")
+async def test_connect_to_an_existing_cdp_session_twice(
+    launch_arguments: Dict, browser_type: BrowserType, server: Server
+):
+    browser_server = await browser_type.launch(
+        **launch_arguments, args=[f"--remote-debugging-port={PORT}"]
+    )
+    endpoint_url = f"http://localhost:{PORT}"
+    cdp_browser1 = await browser_type.connect_over_cdp(endpoint_url)
+    cdp_browser2 = await browser_type.connect_over_cdp(endpoint_url)
+    assert len(cdp_browser1.contexts) == 1
+    page1 = await cdp_browser1.contexts[0].new_page()
+    await page1.goto(server.EMPTY_PAGE)
+
+    assert len(cdp_browser2.contexts) == 1
+    page2 = await cdp_browser2.contexts[0].new_page()
+    await page2.goto(server.EMPTY_PAGE)
+
+    assert len(cdp_browser1.contexts[0].pages) == 2
+    assert len(cdp_browser2.contexts[0].pages) == 2
+
+    await cdp_browser1.close()
+    await cdp_browser2.close()
+    await browser_server.close()
+
+
+def _ws_endpoint_from_url(url: str) -> str:
+    response = requests.get(url)
```

Comment:
Can be done using `urllib.request.urlopen` and then json.loads

Reply:
using requests since its more convenient